flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/RecordContext.java

```diff
@@ -0,0 +1,138 @@
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A context that preserves the necessary variables required by each operation, all operations for
+ * one record will share the same element context.
+ *
+ * <p>Reference counting mechanism, please refer to {@link ContextStateFutureImpl}.
+ *
+ * @param <R> The type of the record that extends {@code
+ *     org.apache.flink.streaming.runtime.streamrecord.StreamElement}.
+ * @param <K> The type of the key inside the record.
+ */
+public class RecordContext<R, K> extends ReferenceCounted {
+
+    /** The record to be processed. */
+    @Nullable private final R record;
+
+    /** The key inside the record. */
+    private final K key;
+
+    /** Whether this Record(Context) has occupied the corresponding key. */
+    private volatile boolean keyOccupied;
+
+    /**
+     * The KeyAccountingUnit that this record context will interact with. The reason why it keeps
+     * this reference here is that it would release the key within {@link
+     * #referenceCountReachedZero()}, which may be called once the ref count reaches zero in any
+     * thread.
+     */
+    private final KeyAccountingUnit<R, K> keyAccountingUnit;
+
+    /**
+     * Whether the record is already in buffer. If yes, this indicates that it is a state request
+     * resulting from a callback statement, otherwise, it signifies the initial state request for a
+     * newly entered record.
+     */
+    private volatile boolean inBuffer;
+
+    /**
+     * The reference of {@link AsyncExecutionController#inFlightRecordNum} that this record context
+     * will interact with. This reference is kept here to update related metrics when a record
+     * context is released.
+     */
+    private final AtomicInteger inflightRecordNumRef;
+
+    RecordContext(
+            KeyAccountingUnit<R, K> keyAccountingUnit,
+            AtomicInteger inflightRecordNumRef,
+            R record,
+            K key) {
+        super(0);
+        this.keyAccountingUnit = keyAccountingUnit;
+        this.inflightRecordNumRef = inflightRecordNumRef;
+        this.record = record;
+        this.key = key;
+        this.keyOccupied = false;
+        this.inBuffer = false;
+    }
+
+    public K getKey() {
+        return this.key;
+    }
+
+    /**
+     * Try to occupy the key via {@link KeyAccountingUnit}. When it has been already occupied once,
+     * the record could continue running. Package-private.
+     *
+     * @return true if occupying succeed.
+     */
+    boolean tryOccupyKey() {
+        return keyOccupied || (keyOccupied = keyAccountingUnit.occupy(record, key));
+    }
+
+    boolean isInBuffer() {
+        return inBuffer;
+    }
+
+    void setInBuffer() {
+        this.inBuffer = true;
+    }
+
+    @Override
+    protected void referenceCountReachedZero() {
+        if (keyOccupied) {
+            keyAccountingUnit.release(record, key);
+            keyOccupied = false;
+            inflightRecordNumRef.decrementAndGet();
```

Comment:
How about only provide a `disposer` to `RecordContext` in constructor, and let `AEC` defines how to dispose a `RecordContext`? It prevents holding `keyAccountingUnit`  as well as `inflightRecordNumRef`.